CMake: Install cmake files into a versioned directory

cjmayo opened the pull request:
Under wxBUILD_INSTALL_LIBRARY_DIR.

---

To allow both wxWidgets 3.2 and 3.3 to be installed at the same time. That seems like it is going to be useful - I tried compiling 2 applications against 3.3 and both failed, although I could force one to complete using `-DwxNO_REQUIRE_LITERAL_MSGIDS`.

CMake can handle this, including finding a specific version e.g.:

```
cmake_minimum_required(VERSION 3.10)
project(test-wx-cmake)
find_package(wxWidgets 3.3 REQUIRED COMPONENTS net core base CONFIG)
message("wxWidgets_VERSION " ${wxWidgets_VERSION})
```

cjmayo committed:
CMake: Install cmake files into a versioned directory

Under wxBUILD_INSTALL_LIBRARY_DIR.

cjmayo committed:
CMake: Fix wx_get_install_dir on MS Windows

On installation:

CMake Error at cmake_install.cmake:3338 (file):
  file cannot create directory: C:/Program Files
  (x86)/wxWidgets/$<PATH:CMAKE_PATH,NORMALIZE,lib$<1:/>clang_x64_dll/cmake/wxWidgets-3.3/clang_x64_dll>.

Also prevents a trailing "/" being added to the wxBUILD_INSTALL_*_DIR on
other platforms.

Suggested-by: Maarten Bent <MaartenBent@users.noreply.github.com>

MaartenBent commented:
I think it fails now because it installs the cmake files in `lib/vc_x64_lib/cmake/wxWidgets-3.3`. It should be in `lib/cmake/wxWidgets-3.3`. Otherwise the default [search rules](https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure) won't find it.

So in this case `wx_get_install_dir` should not append `wxPLATFORM_LIB_DIR`.

Maybe you can temporary disable `wxBUILD_INSTALL_PLATFORM_SUBDIR` so it doesn't get added.
```cmake
set(install_subdir_temp ${wxBUILD_INSTALL_PLATFORM_SUBDIR})
set(wxBUILD_INSTALL_PLATFORM_SUBDIR OFF)
wx_get_install_dir(library "lib")
set(wxBUILD_INSTALL_PLATFORM_SUBDIR ${install_subdir_temp})
```

Or alternatively add a 3th parameter to `wx_get_install_dir` with `TRUE/FALSE` to indicate if it should be added.

cjmayo committed:
CMake: Add a wx_get_build_install_dir macro

On MSW/MSVC with wxMSW .cmake files were being installed into:
C:/Program Files (x86)/wxWidgets/lib/vc_x64_dll/cmake/wxWidgets-3.3/

which is not in the search path.

Add a wx_get_build_install_dir macro which allows a directory to be
retrived without wxPLATFORM_LIB_DIR appeneded, even on platforms that
use it.

vadz commented:
Thinking more about this, I don't know if we're really supposed to do this, shouldn't it be possible to use 3.3 CMake file to find both 3.3 and 3.2? And could this break something relying on finding the files in the directory without version in it? And which version will be used if there is 3.2 version of CMake files in the unversioned directory and 3.3 one in the versioned directory?

On an unrelated note, if you find any unintentional (i.e. not documented on the top of the change log) incompatible changes in 3.3, please open issues for them.

MaartenBent commented:
Yes, in general finding wxWidgets like this should still work fine.

Only existing projects that have configured `wxWidgets_DIR` to point directly to `wxinstalldir/lib/cmake/wxWidgets` will not find it anymore. But normally it should just point to `wxinstalldir` and then it should find the new directory automatically.

I tested what happens when using `find_package(wxWidgets CONFIG)` when there are `wxWidgets`, `wxWidgets-3.2` and `wxWidgets-3.3` and it seems to always find unnumbered first, and then `3.2`. So results are probably ordered alphanumeric, not by latest available version.
When specifying the desired version (`find_package(wxWidgets 3.3 CONFIG)`) it finds the desired one.

So this change should probably be in the changelog. Maybe even as an incompatible change?
Just to inform people that install/overwrite wxWidgets in the same directory, to remove the old non-versioned directory.
